### ISSUE TITLE
Layout panics when there's an unconnected shadow root

### ISSUE THREAD

**Issue by simonwuelker** [servo/servo#34358]
The following triggers a panic in `handle_reflow` when loaded with `--pref dom.shadowdom.enabled`:
```html
<script>
  var foo = document.createElement("div");
  var shadow_root = foo.attachShadow({mode: "open"});
  shadow_root.appendChild(document.createElement("div"));

  var bar = document.createElement("div");
  bar.appendChild(foo);
</script>
```
Layout expects the shadow root to be attached to the document, which it is not. Which begs the question why layout is dealing with it in the first place.

Full backtrace:
```
assertion failed: sr.upcast::<Node>().get_flag(NodeFlags::IS_CONNECTED) (thread Script(1,1), at components/script/layout_dom/document.rs:70)
   0: servoshell::backtrace::print
             at /home/alaska/servo/ports/servoshell/backtrace.rs:18:5
   1: servoshell::panic_hook::panic_hook
             at /home/alaska/servo/ports/servoshell/panic_hook.rs:40:17
   2: core::ops::function::Fn::call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/ops/function.rs:79:5
   3: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/boxed.rs:2077:9
      std::panicking::rust_panic_with_hook
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:799:13
   4: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:656:13
   5: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/sys_common/backtrace.rs:171:18
   6: rust_begin_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:652:5
   7: core::panicking::panic_fmt
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:72:14
   8: core::panicking::panic
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panicking.rs:146:5
   9: script::layout_dom::document::ServoLayoutDocument::shadow_roots::{{closure}}
             at /home/alaska/servo/components/script/layout_dom/document.rs:70:21
  10: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/iter/adapters/map.rs:89:28
  11: <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/slice/iter/macros.rs:230:27
  12: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/iter/adapters/map.rs:129:9
  13: core::iter::traits::iterator::Iterator::for_each
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/iter/traits/iterator.rs:818:9
  14: alloc::vec::Vec<T,A>::extend_trusted
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/vec/mod.rs:3096:17
  15: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/vec/spec_extend.rs:26:9
  16: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/vec/spec_from_iter_nested.rs:62:9
  17: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/vec/spec_from_iter.rs:33:9
  18: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/alloc/src/vec/mod.rs:2970:9
  19: core::iter::traits::iterator::Iterator::collect
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/iter/traits/iterator.rs:2005:9
  20: script::layout_dom::document::ServoLayoutDocument::shadow_roots
             at /home/alaska/servo/components/script/layout_dom/document.rs:66:13
  21: script::layout_dom::document::ServoLayoutDocument::flush_shadow_roots_stylesheets
             at /home/alaska/servo/components/script/layout_dom/document.rs:87:32
  22: layout_thread_2020::LayoutThread::handle_reflow
             at /home/alaska/servo/components/layout_thread_2020/lib.rs:758:9
  23: <layout_thread_2020::LayoutThread as script_layout_interface::Layout>::reflow::{{closure}}
             at /home/alaska/servo/components/layout_thread_2020/lib.rs:507:16
  24: profile_traits::time::profile
             at /home/alaska/servo/components/shared/profile/time.rs:221:9
  25: <layout_thread_2020::LayoutThread as script_layout_interface::Layout>::reflow
             at /home/alaska/servo/components/layout_thread_2020/lib.rs:503:9
  26: script::dom::window::Window::force_reflow
             at /home/alaska/servo/components/script/dom/window.rs:1907:9
  27: script::dom::window::Window::reflow
             at /home/alaska/servo/components/script/dom/window.rs:1982:29
  28: script::dom::document::Document::finish_load
             at /home/alaska/servo/components/script/dom/document.rs:2253:21
  29: script::dom::servoparser::ServoParser::finish
             at /home/alaska/servo/components/script/dom/servoparser/mod.rs:674:9
  30: script::dom::servoparser::ServoParser::do_parse_sync
             at /home/alaska/servo/components/script/dom/servoparser/mod.rs:592:13
  31: script::dom::servoparser::ServoParser::parse_sync
             at /home/alaska/servo/components/script/dom/servoparser/mod.rs:541:35
  32: <script::dom::servoparser::ParserContext as net_traits::FetchResponseListener>::process_response_eof
             at /home/alaska/servo/components/script/dom/servoparser/mod.rs:1003:13
  33: script::script_thread::ScriptThread::handle_fetch_eof
             at /home/alaska/servo/components/script/script_thread.rs:4201:13
  34: script::script_thread::ScriptThread::handle_msg_from_constellation
             at /home/alaska/servo/components/script/script_thread.rs:2328:25
  35: script::script_thread::ScriptThread::handle_msgs::{{closure}}
             at /home/alaska/servo/components/script/script_thread.rs:1949:25
  36: script::script_thread::ScriptThread::profile_event
             at /home/alaska/servo/components/script/script_thread.rs:2293:13
  37: script::script_thread::ScriptThread::handle_msgs
             at /home/alaska/servo/components/script/script_thread.rs:1942:26
  38: script::script_thread::ScriptThread::start
             at /home/alaska/servo/components/script/script_thread.rs:1393:15
  39: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}::{{closure}}
             at /home/alaska/servo/components/script/script_thread.rs:863:25
  40: profile_traits::mem::ProfilerChan::run_with_memory_reporting
             at /home/alaska/servo/components/shared/profile/mem.rs:91:9
  41: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}
             at /home/alaska/servo/components/script/script_thread.rs:861:17
  42: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/sys_common/backtrace.rs:155:18
  43: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/thread/mod.rs:542:17
  44: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/panic/unwind_safe.rs:272:9
  45: std::panicking::try::do_call
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:559:40
  46: __rust_try
  47: std::panicking::try
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panicking.rs:523:19
  48: std::panic::catch_unwind
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/panic.rs:149:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/std/src/thread/mod.rs:541:30
```

**Comment by simonwuelker** [comment]
The issue is that all shadow roots in `document.shadow_roots` are expected to be connected, but `Document::register_shadowroot` doesn't enforce that.
https://github.com/servo/servo/blob/5bd28f2abda5f4ca87dac852d2bec82f5fce0539/components/script/dom/document.rs#L4038-L4043

Now, who is in the wrong here?

**Comment by simonwuelker** [comment]
I think turning that assertion into a `.filter` call should be fine. I can't think of a reason layout would have to worry about unconnected shadow roots.

**Comment by jdm** [comment]
I guess the question is whether there's any need for a document to be aware of shadow roots that aren't connected? We could add and remove them from the list based on the appropriate VirtualMethod calls.

**Comment by jdm** [comment]
This was fixed by #34358.
